### ISSUE TITLE
Feature planning and discussion

### ISSUE THREAD

**Issue by VShawn** [VShawn/PRemoteM#26]
- [x] Port from .NET Framework to .NET Core (#72)
- [ ] Default username/password for protocol and other attributes ? (#73)
- [x] Bulk edit all attributes (#83)
- [x] Refactor keyword match model (#74)
- [x] protocol
  - [x] RdpApp like https://github.com/kimmknight/remoteapptool (#60)
  - [x] protocol settings template  (#22)
  - ~add spice support?~
  - [x] external tool embed support? (#61)
  - [x] sFTP/FTP with in SSH?
- [x] Chrome like tab window title #28
   - [x] design a good look UI
   - [x] implement change
- [x] UI
  - [x] better experience for server management 
       - [x] right click to edit https://github.com/VShawn/PRemoteM/issues/22#issuecomment-740849283 
       - [x] list manager
       - [x] editable textblock for list manager to edit base info
       - [x] change multi-server info on one operation(i.e. change addree). #22
  - [x] list view for home page.
  - [x] orderby options refactor 
  - [x] Typing in main UI should go to search #34
  - [x] close button on tab header. #34
  - [x] Remember window maximization state
  - [x] save server icon without save button
- [ ]  Sync database. #76

**Comment by majkinetor** [comment]
>  sFTP/FTP with in SSH?

You could perhaps integrate WinSCP ?

**Comment by majkinetor** [comment]
Launcher items could bi bigger, second line is unreadable and RDP is barely visible

<details><summary>screenshot</summary> 

![image](https://user-images.githubusercontent.com/85767/101528657-df09cc80-398f-11eb-9212-c6ce9b8425cc.png)
</details>

**Comment by VShawn** [comment]
# 1

> Launcher items could bi bigger, second line is unreadable and RDP is barely visible
> 
> screenshot

it seens that you are using a 2k screen? 

now font size is ok for me with 1080P screen or 4k 200%  scale screen.

![image](https://user-images.githubusercontent.com/10143738/101569391-5e8ab080-3a0f-11eb-9fae-d88467627383.png)

maybe it is better to offering some options to customize the font-size 👀 

# 2

> > sFTP/FTP with in SSH?
> You could perhaps integrate WinSCP ?

i not attempt to integrate WinSCP now, WinSCP is powerful and complicate, I consider to add feature like mRemoteNG which allow integrate external apps: see https://mremoteng.readthedocs.io/en/latest/howtos/external_tools.html

for these above more investigations are needed to check what is support by command line for winscp, it will take a lot of time, and i don't have much time 😭 
> .\winscp.exe sftp://user:123@127.0.0.1:22

so the main target for PRemoteM now is to offering a simple and easy way to access remote session and make it a stable version, after that we will refact PRemoteM and add some add powerful features for it.

as you can see, there is even no time for me to build a unit test project 😞

**Comment by majkinetor** [comment]
1. I am on 1980x1080 screen. Perhaps add font/scale options on planned UI refactoring?
2. You integrated putty, winscp is has Net dll, powershell module etc and looks 100% programmable. Just mentioning it in case you were not aware.

> as you can see, there is even no time for me to build a unit test project

I am not sure this project requires it - better to have more users IMO, which will get bug reports up. I am deeply involved in automatic tests of all kinds, and I personally don't see a big benefit (maybe for few stuff like import/export only). I plan to publish chocolatey package if I continue to use it after trial which might bring more users. I will also report bugs and maybe some ideas as long as you listen :)

**Comment by VShawn** [comment]
@majkinetor 👍 
thank you, bug reports and new ideas are always welcome, I will try my best to fix them. 
have a nice day.

**Comment by majkinetor** [comment]
FYI, https://winscp.net/eng/docs/library_install

**Comment by majkinetor** [comment]
Doesn't look it has core/5 version tho...

**Comment by valsamis-d** [comment]
Perhaps support SSH tunneling via SOCKS proxy?

**Comment by VShawn** [comment]
> Perhaps support SSH tunneling via SOCKS proxy?

I am thinking about this, still looking for how to support proxy.

**Comment by bluikko** [comment]
I would suggest to automatically change the port to the protocol default port when changing protocol type tab. Maybe only if the port was previously the default port and not a custom port.

For example:
- RDP -> set port to 3389
- SSH -> set port to 22
etc.

Second suggestion I have is that when changing focus into a textbox control with the `tab` button, automatically select all the current contents in the textbox.

**Comment by majkinetor** [comment]
@bluikko , you told on mRemoteNg repo that some features are missing? Care to elaborate ?

**Comment by bluikko** [comment]
Compare the settings per host in mRemoteNG and PRemoteM. mRemoteNG allows to change about every setting of RDP/SSH (have not used other functionality). Many of those are useful.

It also provides inheritance from groups to group members.

**Comment by VShawn** [comment]
> Compare the settings per host in mRemoteNG and PRemoteM. mRemoteNG allows to change about every setting of RDP/SSH (have not used other functionality). Many of those are useful.

I'd like to keep setting to be simple and easy, or minimal configuration to go. So I make RDP setting reference to `mstsc.exe`, not reference to `mRemoteNG`. besides I have no idea about the meaning of some setting in mRemoteNG RDP(Load balance Info, VM ID...) If you find it really useful, just tell me and I will add the setting.

BTW we have almost the same settings in SSH with mRemoteNG
![image](https://user-images.githubusercontent.com/10143738/106088536-1e1ef780-6161-11eb-9080-677eeef552a6.png)



> inheritance

inheritance will not be implemented in PRM for now, since we offering a simple multiple edit in management page. a better UX will be offered in future release.

**Comment by majkinetor** [comment]
> It also provides inheritance from groups to group members.

Which as far as I can tell doesn't work.  mRemoteNG interface for setting connection is IMO very bad design. It doesn't allow for fast and fluid experience. Everything is also much more slower then with PRM.

> BTW we have almost the same settings in SSH with mRemoteNG

This is my impression too. I have around 50 connections of all types that I moved from mRemoteNg and I didn't miss anything in particular. Furthermore, major deal breaker was focus problem and single monitor problem.

While PRM certainly needs some more work, its already great tool that makes you productive without going in your way.

**Comment by bluikko** [comment]
I am sure it is a matter of tastes and use cases. I would love to get rid of mRemoteNG since it has major bugs but so far I've not found another open-source connection manager that can do what I need.

For me for example: it is a complete no-starter that putty profiles cannot (?) be defined per-connection.

Does it work with Pageant or other SSH agent? Selecting private key from filesystem does not work with a CAC.

**Comment by majkinetor** [comment]
> it is a complete no-starter that putty profiles cannot (?) be defined per-connection.

Yeah, this is something that should definitely be implemented. I added a ticket #68

**Comment by VShawn** [comment]
> Does it work with Pageant or other SSH agent? Selecting private key from filesystem does not work with a CAC.

No it dont, I dont know what Pageant or a CAC is, never use it before. You can teach me how it works.....

**Comment by majkinetor** [comment]
It remembers the passphrase for ssh for some time key so that you don't have to enter it each time.  It is not used with password access, only with key access. This is the main feature for people who use keys (and most use keys, since putty didn't allow for password automation).

Don't know about CAC.

**Comment by bluikko** [comment]
Common Access Cards are used in the defense industry. Basically smart cards.

They look like a regular key pair when used with the patched Pageant from PuTTY-CAC.

**Comment by Wand3rvogel** [comment]
Support for Windows Server 2016 RDS will be nice.

**Comment by adelphi-company** [comment]
I'd love to see the possibility to connect to Hyper-V guests consoles using their VM ID. We often use this in RDCman and mRemoteNG supports it too.

We also like to see better multi user support as we share our connection configs within our team. Therefore it would be nice to have a global username / password / key setting that will be used if no credentials are provided within a connection profile. Supporting windows environmental variables like %USERNAME% would be another way to solve this.

**Comment by VShawn** [comment]
@Wand3rvogel 

> Support for Windows Server 2016 RDS will be nice.

Sorry I can‘t develop this feature until I can setup( or somebody help me to setup) a RDS environment for test (I have tried to set a RDS up, but failed.)

---

@adelphi-company 

I did have the plan to support `Hyper-V guests consoles using their VM ID` and `better multi user support`. But because I don’t have enough time, I have been delayed until now. Now I‘m working on tag system for bulk connect #107. And after #107 is done, I will try to make the improvement for `better multi user support`.

**Comment by majkinetor** [comment]
> We also like to see better multi user support as we share our connection configs within our team. Therefore it would be nice to have a global username / password / key setting that will be used if no credentials are provided within a connection profile. Supporting windows environmental variables like %USERNAME% would be another way to solve this.

I created ticket for this #128

**Comment by stetze88** [comment]
> Support for Windows Server 2016 RDS will be nice.
Yes. 👍🏻 

We need an field for the loadbalanceinfo:s:tsv://MS Terminal Services Plugin.1.<c_collection>

- loadbalanceinfo is the parameter name
- s is the type of value that this parameter expects. "S" means "String". I'm sure that if you look in your RDP file, you will find settings with "i", it means "Integer".
- tsv://MS Terminal Services Plugin.1.<c_collection> is the value of the parameter "loadbalanceinfo", it's a string.
[https://serverfault.com/questions/913236/need-explanation-of-tag-loadbalanceinfostsv](url)

**Comment by VShawn** [comment]
@stetze88 
Yes I found the `loadbalanceinfo` property:

https://docs.microsoft.com/en-us/windows/win32/termserv/imsrdpclientadvancedsettings-loadbalanceinfo

which means it is easy to add the field into the app. I need find somewhere to place this `low usage` field, and I want you to help me test if this field works.

**Comment by stetze88** [comment]
@VShawn 
I'm happy to check the function of the field.

**Comment by vNa3006** [comment]
I think a password to open a certain database would be good. Or a password required to open the application. So not everyone has access to my computer can just open PRemoteM and access to the servers I manage.

**Comment by vNa3006** [comment]
> @vNa3006 THX for advising, I may not support password protect for PRemotM right now. see https://github.com/VShawn/PRemoteM/wiki/Security

People usually overthinking about simple stuff.
I personally don't use any app that doesn't support password protection. It's easy and take few minutes to build too.

Anyway there are reasons to use and not to use it. Thanks for sharing.

**Comment by majkinetor** [comment]
> I personally don't use any app that doesn't support password protection.

That is amazingly inconvenient life. 

I agree here with @VShawn and his wiki notes. He did skip one important method, the one utilized by some tools like sudo, putty agent etc - i.e. unlocking a system for certain amount of time which is configurable (such as 15 minutes). 

However, with physical access to your machine, all that is void. Outside of machine (e.g. if you keep it on git repo or NAS share or whatever) you can protect it via other means.

**Comment by vNa3006** [comment]
@majkinetor it's OK. You don't have to share your computer with anyone.
There are millions, if not billions of people have to share their computer with someone else.

Anyway keep it simple. I don't use PRemoteM because it doesn't have password protection. That's all.

**Comment by majkinetor** [comment]
>Anyway keep it simple

We are :) Thanks for passing buy and letting us know.

**Comment by vNa3006** [comment]
Lol, it's OK. It's good to know we don't have that much human beings on the Earth.

**Comment by VShawn** [comment]
> I don't use PRemoteM because it doesn't have password protection. That's all.

Sorry to hear that.

> There are millions, if not billions of people have to share their computer with someone else.

For me, i will choose the guest account if I have to share my computer with someone else. And of course the choose can not access my sensitive data. A app lelvel password is inessential in this case.

If I have no choose but share computer with someone else in a same account(or the os is not on my control), then I prefer a portable app with [Encrypted USB Flash Drives](https://www.westerndigital.com/en-au/products/usb-flash-drives/sandisk-cruzer-glide-usb-2-0#SDCZ60-016G-BQ35) which can be protected by password or tourch id.

https://www.amazon.com/biometric-flash-drive/s?k=biometric+flash+drive

**Comment by vNa3006** [comment]
@VShawn thanks for your suggestion.
Right now I just use mRemoteNG and have different passwords for different configuration files.

**Comment by majkinetor** [comment]
> @majkinetor, I have to admit we did have to share machine with workmates somethime: like we have a machine-learning workstation that shared to all of the AI engineers.

Yeah, I know, there are all kind of dumb things in the world. I consulted many companies and they do that kind of thing. You probably shouldn't work for such lax security companies as it may backfire legally on you.

Any machine can be shared by having personalized users which is how OSes work for more then half a century. 

> then I prefer a portable app

Yes, that is one solution. @vNa3006 can keep all his data on the USB drive and never leave without it.

**Comment by majkinetor** [comment]
When I say access to computer, I really mean access to the computer using **the same user**.

**Comment by grbd** [comment]
One thing I've been trying to get is a RDP session
where the remote resolution is 1920x1080
and local resolution is 3840x2160
and the zoom is set to 200% so that it works full screen
(this is to avoid issues with other users connecting to the same box and scaling issues with icons)

So far I've only found the "Microsoft Remote Desktop App" on the windows store that does that
(both zooming and fullscreen at the same time), but the ui on that one is pretty awful compared to anything else

That would be a nice feature to have, but it sounds like you're limited by the ActiveX control

**Comment by VShawn** [comment]
@grbd 

I dot understand, if your loacal resolution is 3840x2160, why your remote full-screen is 1920x1080?

IMO, your remote full-screen would be 3840x2160 as well.

BTW, I just fell it unnecessary to let user set the zooming manually, PRemoteM will automatically set the zooming based on Windows display settings.

![image](https://user-images.githubusercontent.com/10143738/145920685-da35a323-32b2-49ce-91f8-97c634e55395.png)

**Comment by grbd** [comment]
The general gist is when I connect to a box under a generic admin account

  * I set the resolution of the remote side to something like 1920x1080
  * The resolution of the local side is 3840x2160
  * With the option to zoom or stretch the window so it's fully visible under 3840x2160

The reasons for doing it this way include

  * it avoids screwing up the desktop icons / window positions for other users that might logon using the same username that don't have a 4K desktop.
  * In some cases the windows boxes I'm logging onto might be old and don't have a dpi option which results in tiny icons on a UHD 4K laptop screen.

**Comment by VShawn** [comment]
Most of the points in this ticket have already been completed
